### PR TITLE
Add jsonnet lib + tests and refactor logic for caching configs

### PR DESCRIPTION
**Pull request by saswatamcode**
This PR does the following,

- Refactors main logic with two interfaces, `tenantRulesLoader` and `tenantRulesSyncer`, and adds a function which runs the main loop `SyncLoop` using these. This makes functionality easier to test and caches obsctl credentials locally so that we don't continuously make requests to SSO for new tokens.
- Add unit tests for core logic
- Adds jsonnet library to help in easier deployments in rhobs/configuration. Also, the openshift template is now generated via the lib
- Add CI jobs and `bingo` tooling

The reason for this effort is that we plan to use `obsctl-reloader` to generate synthetic data for measuring RHOBS Rules API SLO. `obsctl-reloader` can help in maintaining a consistent stream of traffic to the required Observatorium endpoints. 🙂 

Relevant ticket: https://issues.redhat.com/browse/MON-2735

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -13,35 +13,45 @@ import (
...
 
-const obsctlContext = "api"
+const obsctlContextAPIName = "api"
 const defaultSleepDurationSeconds = 30
 
-var logger log.Logger
+// tenantRulesLoader represents logic for loading and filtering PrometheusRule objects by tenants.
+// Useful for testing without spinning up cluster.
+type tenantRulesLoader interface {
+	GetPrometheusRules() ([]*monitoringv1.PrometheusRule, error)
```
Comment (jessicalins) [inline main.go:30]: do `GetPrometheusRules` and `GetTenantRuleGroups` need to be exported? Looks like they are being used only in `main` pkg?

Comment (jessicalins) [reply main.go:30]: same for the other funcs in this file (e.g. `InitOrReloadObsctlConfig`, `SetCurrentTenant`, etc)

**Review by bwplotka — APPROVED** [review]
Amazing! Thanks for tests 💪🏽

File: main.go [hunk 2]
```diff
@@ -155,26 +232,29 @@ func main() {
...
+		level.Error(logger).Log("msg", "error reloading/initializing obsctl config", "error", err)
+		os.Exit(1)
+	}
+
+	sleepDurationSeconds := defaultSleepDurationSeconds
+	if value, ok := os.LookupEnv("SLEEP_DURATION_SECONDS"); ok {
+		sleepDurationSeconds, _ = strconv.Atoi(value)
+	}
+
+	SyncLoop(ctx, logger,
```
Comment (jessicalins) [inline main.go:253]: I think this can be unexported as well

File: main_test.go [hunk 3]
```diff
@@ -0,0 +1,434 @@
...
+	}
+
+	for _, tc := range []struct {
+		name             string
+		tenants          string
+		prexistingConfig *config.Config
+		wantConfig       *config.Config
+	}{
+		{
+			name: "already existing config with one tenant",
```
Comment (jessicalins) [inline main_test.go:82]: I wonder if it makes sense to test with an empty config as well?

File: main_test.go [hunk 4]
```diff
@@ -0,0 +1,434 @@
...
+									Expr:  intstr.FromString("vector(1)"),
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+		{
+			name:    "multiple tenant with multiple rulegroups",
```
Comment (jessicalins) [inline main_test.go:331]:
```suggestion
			name:    "multiple tenants with multiple rulegroups",
```

**Review by jessicalins — COMMENTED** [review]
awesome refactor! :sparkles: I left some comments/suggestions

File: .bingo/README.md [hunk 5]
```diff
@@ -0,0 +1,14 @@
+# Project Development Dependencies.
+
+This is directory which stores Go modules with pinned buildable package that is used within this repository, managed by https://github.com/bwplotka/bingo.
```
[the remark this answers is not in the document]
Comment (saswatamcode) [reply .bingo/README.md:3]: This is auto-generated by bingo! 🙂

File: Makefile [hunk 6]
```diff
@@ -1,3 +1,39 @@
...
+
+# Tools.
+GIT ?= $(shell which git)
+
+# Support gsed on OSX (installed via brew), falling back to sed. On Linux
+# systems gsed won't be installed, so will use sed as expected.
+SED ?= $(shell which gsed 2>/dev/null || which sed)
+
+define require_clean_work_tree
+	@git update-index -q --ignore-submodules --refresh
```
[the remark this answers is not in the document]
Comment (saswatamcode) [reply Makefile:19]: Yup, nice catch!

File: main.go (hunk 1 above)
Comment (saswatamcode) [reply main.go:30]: Will change!

File: main_test.go (hunk 3 above)
Comment (saswatamcode) [reply main_test.go:82]: Will add a test case!

**Review by jessicalins — APPROVED** [review]
:1st_place_medal: cool!

File: .bingo/README.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (jessicalins) [reply .bingo/README.md:3]: ah right!

**Review by maorfr — APPROVED** [review]

**Pull request merged**